syntax = "proto3";

package protos;

option go_package = "github.com/Proxmox-Cloud/terraform-provider-pxc/internal/provider/protos;protos";

service CloudService {
  rpc GetMasterKubeconfig(GetKubeconfigRequest) returns (GetKubeconfigResponse);
  rpc GetClusterVars(GetClusterVarsRequest) returns (GetClusterVarsResponse);
  rpc GetCloudFileSecret(GetCloudFileSecretRequest) returns (GetCloudFileSecretResponse);
  rpc CreateCloudSecret(CreateCloudSecretRequest) returns (CreateCloudSecretResponse);
  rpc DeleteCloudSecret(DeleteCloudSecretRequest) returns (DeleteCloudSecretResponse);
  rpc GetCloudSecret(GetCloudSecretRequest) returns (GetCloudSecretResponse);
  rpc GetCloudSecrets(GetCloudSecretsRequest) returns (GetCloudSecretsResponse);
  rpc GetCephAccess(GetCephAccessRequest) returns (GetCephAccessResponse);
  rpc GetSshKey(GetSshKeyRequest) returns (GetSshKeyResponse);
  rpc GetProxmoxApi(GetProxmoxApiRequest) returns (GetProxmoxApiResponse);
  rpc CreateProxmoxApi(CreateProxmoxApiRequest) returns (CreateProxmoxApiResponse);
  rpc DeleteProxmoxApi(DeleteProxmoxApiRequest) returns (DeleteProxmoxApiResponse);
  rpc GetProxmoxHost(GetProxmoxHostRequest) returns (GetProxmoxHostResponse);
  rpc GetPveInventory(GetPveInventoryRequest) returns (GetPveInventoryResponse);
  rpc GetVmVarsBlake(GetVmVarsBlakeRequest) returns (GetVmVarsBlakeResponse);
}

message GetPveInventoryRequest {
  string target_pve = 1;
}

message GetPveInventoryResponse {
  string inventory = 1;
  string cloud_domain = 2;
}

message GetProxmoxHostRequest {
  string target_pve = 1;
}

message GetProxmoxHostResponse {
  string pve_host = 1;
}

message GetProxmoxApiRequest {
  string target_pve = 1;
  string api_path = 2;
  map<string, string> get_args = 3;
}

message GetProxmoxApiResponse {
  string json_resp = 1;
}

message CreateProxmoxApiRequest {
  string target_pve = 1;
  string api_path = 2;
  map<string, string> create_args = 3;
}

message CreateProxmoxApiResponse {
  bool success = 1;
  string err_message = 2;
}

message DeleteProxmoxApiRequest {
  string target_pve = 1;
  string api_path = 2;
}

message DeleteProxmoxApiResponse {
  bool success = 1;
  string err_message = 2;
}

message GetSshKeyRequest {
  string target_pve = 1;
  enum KeyType {
    AUTOMATION = 0;
    PVE_HOST_RSA = 1;
  }
  KeyType key_type = 2;
}

message GetSshKeyResponse {
  string key = 1;
}

message GetCephAccessRequest {
  string target_pve = 1;
}

message GetCephAccessResponse {
  string ceph_conf = 1;
  string admin_keyring = 2;
}

message GetKubeconfigRequest {
  string target_pve = 1;
  string stack_name = 2;
}

message GetKubeconfigResponse {
  string config = 1;
}

message GetClusterVarsRequest {
  string target_pve = 1;
}

message GetClusterVarsResponse {
  string vars = 1;
}

message GetCloudFileSecretRequest {
  string target_pve = 1;
  string secret_name = 2;
  bool rstrip = 3;
}

message GetCloudFileSecretResponse {
  string secret = 1;
}

message CreateCloudSecretRequest {
  string target_pve = 1;
  string secret_name = 2;
  string secret_data = 3;
  string secret_type = 4;
}

message CreateCloudSecretResponse {
  bool success = 1;
  string err_message = 2;
}

message DeleteCloudSecretRequest {
  string target_pve = 1;
  string secret_name = 2;
}

message DeleteCloudSecretResponse {
  bool success = 1;
  string err_message = 2;
}

message GetCloudSecretRequest {
  string target_pve = 1;
  string secret_name = 2;
}

message GetCloudSecretResponse {
  string secret = 1;
}

message GetCloudSecretsRequest {
  string target_pve = 1;
  string secret_type = 2;
}

message GetCloudSecretsResponse {
  string secrets = 1;
}

message GetVmVarsBlakeRequest {
  string target_pve = 1;
  repeated string blake_ids = 2;
}

message GetVmVarsBlakeResponse {
  map<string, string> blake_id_vars = 1;
}